.github/workflows/test-browserstack.yml

```diff
@@ -0,0 +1,114 @@
+name: BrowserStack Tests
+
+on:
+  pull_request_target:
+    branches:
+      - 'main'
+      - 'v3.0.0-dev'
+  push:
+    branches:
+      - 'main'
+      - 'v3.0.0-dev'
+
+env:
+  BROWSERSTACK_ACCESS_KEY: ${{ secrets.BROWSER_STACK_ACCESS_KEY }}
+  BROWSERSTACK_USERNAME: ${{ secrets.BROWSER_STACK_USERNAME }}
+
+permissions: read-all
+
+# this workflow splits running Browserstack tests into two jobs:
+# - one to build stencil for Browserstack
+# - one to build the test applications & invoke Browserstack
+#
+# doing so allows us to balance a few different considerations when running against Browserstack:
+# 1. Rerunning Failed Jobs
+#    Creating a separate job for running Browserstack on its own allows us to 'rerun failed jobs' in the GitHub UI,
+#    only for the Browserstack tests themselves.
+#    This allows for a quicker feedback loop in the event of flaky jobs
+# 2. Secret Security
+#    In order to authenticate against Browserstack, an access key and username are required
+#    In order to minimize their access/usage elsewhere, we contain them to the workflow that is this file
+#
+# These items do come at a cost of some repeated steps in the jobs below (checking out code, installing dependencies,
+# etc.)
+
+jobs:
+  build_for_browserstack:
+    name: Build for Browserstack
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@ec3a7ce113134d7a93b817d10a8272cb61118579 # v2.4.0
+        with:
+          ref: ${{ github.event_name == 'pull_request_target' && format('refs/pull/{0}/merge', github.event.number) || '' }}
+          persist-credentials: false
+
+      - uses: actions/setup-node@56337c425554a6be30cdef71bf441f15be286854 # v3.1.1
+        with:
+          node-version: '12'
+          cache : 'npm'
+
+      - name: Install dependencies
+        run: npm ci
+
+      - name: Build
+        run: npm run build -- --ci
+
+      - name: Upload Build Artifacts
+        uses: ./.github/workflows/actions/upload-archive
+        with:
+          name: stencil-core
+          output: stencil-core-browserstack-build.zip
+          paths: build cli compiler dependencies.json dev-server internal mock-doc scripts screenshot sys testing
```

Comment:
I experimented with using callable workflow for the second job where we run browserstack. It got a little funky getting the secrets made available in a callable/reusable workflow, in that I couldn't pull them right out of the 'secret context' the GH provides. Instead, you have to pull them in to the caller workflow, then pass them into the called workflow (there are a few ways to do this that make it a little less painful with a 'secrets: inherit` option). That started to feel a little over-engineered, since we could achieve the same result of re-triggering just Browserstack with 2 jobs in the same workflow.

But...I don't know how well plugging `build.yml` in will work here with the current design. I think I played with it a bit, but don't remember exactly remember if/why I would have decided against it.  Let me give that a shot and I'll let you know!

Reply:
If you want to stop fiddling with this stuff and merge as is too feel free to do so :)

Reply:
Nah - where's the fun in that?? 🤣